keps/sig-node/277-ephemeral-containers/README.md

```diff
@@ -206,29 +212,114 @@ There are no limits on the number of Ephemeral Containers that can be created in
 a pod, but exceeding a pod's resource allocation may cause the pod to be
 evicted.
 
-### Restarting and Reattaching Ephemeral Containers
+### Identifying Pods with Ephemeral Containers
 
-Ephemeral Containers will not be restarted.
+The kubelet will set a `PodCondition` when it starts an Ephemeral Container.
+This condition may not be cleared: it will exist for the lifetime of the Pod
+and continues to exist even if all Ephemeral Containers are removed.
 
-We want to be more user friendly by allowing re-use of the name of an exited
-ephemeral container, but this will be left for a future improvement.
+The intended use of this `PodCondition` is to enable administrators to enforce
+custom policies for pods that have had Ephemeral Containers. For example,
+cluster administrators may want to automatically apply a label or delete the pod
+after a configurable time. This may be accomplished by a controller watching
+for this `PodCondition`, though the implementation of such a controller is out
+of scope for this proposal.
+
+### Reattaching and Restarting Ephemeral Containers
 
 One can reattach to a Ephemeral Container using `kubectl attach`. When supported
 by a runtime, multiple clients can attach to a single debug container and share
 the terminal. This is supported by Docker.
 
-### Killing Ephemeral Containers
+Ephemeral Containers will not be restarted automatically, and there is no
+method in the API to restart an Ephemeral Container.  Creators of Ephemeral
+Containers are expected to choose a new, unused name.
+
+### Configurable Security Policy
+
+The ability to add containers to a pod implies security trade offs. We've
+received the following requirements and feedback on the alpha implementation:
+
+*   Admission controllers should be able to enforce policy based on the
+    cumulative pod specification, so operations that prune information,
+    such as removing Ephemeral Containers should not be allowed.
+*   Restarting a pod is disruptive, so for reasons of operation, security,
+    and resource hygiene it should be possible to delete Ephemeral Containers
+    via the API. ([#84764]).
+*   Ephemeral Containers could allow privilege escalation greater than that
+    of the initial pod, so setting a custom security context should not be
+    allowed.
```

Comment:
The main argument against using the existing containers field at the time is that it's (mostly) immutable, and we'd like it to stay that way.  There was concern that a mutable container list will increase the complexity of the kubelet state machine. I also think that the immutable container list as a policy guides users to use pods as intended.

I understand the concern here, though, especially since we're going to be relying on these 3rd party controllers instead of built-ins like PSP going forward. It's a bit of a chicken-and-egg problem since we won't really get the attention of controller developers until it's out of beta and in use in real clusters.

Since we're planning to beta in 1.21, one thing we could do is provide a 2 release buffer where we ship with the feature enabled by default, but disabled by default RBAC policy. That would make it easy for cluster admins to enable on the hosted platform while giving controller developers time to catch up.

My concern with this approach is that I don't know enough about RBAC policy to know if this sort of policy changing would make upgrades/downgrades difficult to manage. @tallclair what do you think?

Reply:
Perhaps another consideration is that third party admission controllers using webhooks will receive a `v1.EphemeralContainers` rather than a `v1.Pod` (when configured for `pods/ephemeralcontainers`). If I were writing a policy controller I would deny this unexpected condition.

Reply:
Added a section below.